sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -369,6 +375,20 @@ class QueryExecution(
     Utils.redact(sparkSession.sessionState.conf.stringRedactionPattern, message)
   }
 
+  /**
+   * This method performs a pre-order traversal and return a boolean Array
+   * representing whether some nodes of the logical tree are persisted.
+   */
+  def computeCacheStateSignature(): Array[Boolean] = {
```

Comment:
How about using `BitSet` for persistence state representation?
It'll be easier to work with and it's more efficient.

Reply:
`BitSet` requires a `numBits` parameter. I cannot know the number of children in advance. Although current implementation is less efficient, it's still acceptable.

Reply:
It's not necessary to know the number of bits during construction.
It's up to you, but just FYI, here is how it'd look like:
```scala
    val builder = BitSet.newBuilder
    var index = 0
    normalized.foreach { fragment =>
      val cached = fragment match {
        case _: IgnoreCachedData => false
        case _ => cacheManager.lookupCachedData(fragment).isDefined
      }
      if (cached) builder += index
      index += 1
    }
    builder.result()
```

Reply:
It seems that scala's `BitSet` requires that the length of elements is divisible by 64.